### PR TITLE
Fix IPC default named pipe clean up on exit

### PR DESCRIPTION
**Pull request by PaulHigin**
# PR Summary

This PR addresses #11599.  On Linux platforms, the default named pipe file was not getting cleaned up on PowerShell exit.

## PR Context

Windows PowerShell uses the CLR Appdomain unload event to dispose the NamedPipeServerStream object on exit, but that does not work with .NET core runtime.  This was not a problem on Windows platforms because the Windows OS releases all process resources on process exit, including named pipes.

Fix is to change this to use the CLR 'ProcessExit' event.

Note that this only works when PowerShell exits normally, i.e., via the 'exit' command.  If the console or terminal window is closed then the CLR cannot fire the event.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs [hunk 1]
```diff
@@ -781,29 +787,60 @@ public string MainWindowTitle
...
+        internal PSHostProcessInfo(
+            string processName,
+            int processId,
+            string appDomainName,
+            string pipeNameFilePath)
         {
-            if (string.IsNullOrEmpty(processName)) { throw new PSArgumentNullException("processName"); }
+            if (string.IsNullOrEmpty(processName))
+            {
+                throw new PSArgumentNullException("processName");
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs]:
```suggestion
                throw new PSArgumentNullException(nameof(processName));
```

File: src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs [hunk 2]
```diff
@@ -781,29 +787,60 @@ public string MainWindowTitle
...
+            }
 
-            if (string.IsNullOrEmpty(appDomainName)) { throw new PSArgumentNullException("appDomainName"); }
+            if (string.IsNullOrEmpty(appDomainName))
+            {
+                throw new PSArgumentNullException("appDomainName");
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs]:
```suggestion
                throw new PSArgumentNullException(nameof(appDomainName));
```

File: src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs [hunk 3]
```diff
@@ -781,29 +787,60 @@ public string MainWindowTitle
...
             try
             {
                 var proc = Process.GetProcessById(processId);
                 MainWindowTitle = proc.MainWindowTitle ?? string.Empty;
             }
-            catch (ArgumentException) { }
-            catch (InvalidOperationException) { }
+            catch (ArgumentException)
+            {
+                // Ignore.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs]: Can we have more useful comment?

File: src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs [hunk 4]
```diff
@@ -781,29 +787,60 @@ public string MainWindowTitle
...
+            }
+            catch (InvalidOperationException)
+            {
+                // Ignore.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs]: The same.

File: src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs [hunk 5]
```diff
@@ -961,30 +959,31 @@ internal static void CreateIPCNamedPipeServerSingleton()
...
-                        {
-                            // Don't throw an exception on the app domain unload event thread.
-                        }
+                        // Terminate the IPC thread.
+                        namedPipeServer.Dispose();
                     }
-                };
+                    catch (ObjectDisposedException)
+                    {
+                        // Ignore.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs]: The same.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Get-PSHostProcessInfo.Tests.ps1 [hunk 6]
```diff
@@ -53,4 +53,49 @@ Describe "Get-PSHostProcessInfo tests" -Tag CI {
...
+                Start-Sleep -Milliseconds 500
+            }
+
+            exit
+'@ | Out-File -FilePath $testfilePath
+
+        # Create PowerShell process to monitor.
+        $psFileName = $IsWindows ? 'pwsh.exe' : 'pwsh'
+        $psPath = Join-Path -Path $PSHOME -ChildPath $psFileName
+        $psProc = Start-Process -FilePath $psPath -ArgumentList "-File $testfilePath -LiveFilePath $aliveFile" -PassThru
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Get-PSHostProcessInfo.Tests.ps1]: We use `Start-Process "$PSHOME/pwsh"` pattern in all tests.

**Review by iSazonov — COMMENTED** [review]
> For Windows PowerShell we were using the CLR Appdomain unload event to dispose of the server pipe, but that is not implemented in dotNet core.

Is this a fundamental Unix behavior/problem or we should open an issue in .Net repo to request a fix?

File: src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs [hunk 7]
```diff
@@ -707,7 +707,7 @@ internal static IReadOnlyCollection<PSHostProcessInfo> GetAppDomainNamesFromProc
                                     else if (process.ProcessName.Equals(pName, StringComparison.Ordinal))
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs]: @PaulHigin, your last commit had 1 failures in `PowerShell-CI-linux`
Basic SysLog tests on Linux.Verifies logging level filtering works
```powershell
Expected $null, but got PSLogItem.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 256
256:         $items | Should -Be $null
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs]: @PaulHigin, your last commit had 1 failures in `PowerShell-CI-macos`
Test-Connection.TraceRoute.TraceRoute works
```powershell
Expected exactly '10.79.0.125', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.165.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 257
257:             $result[0].TargetAddress | Should -BeExactly $realAddress
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs]: @PaulHigin, your last commit had 1 failures in `PowerShell-CI-windows`
ForEach-Object -Parallel Functional Tests.Verifies parallel task queuing
```powershell
Expected the actual value to be greater than 5, but got 5.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\Foreach-Object-Parallel.Tests.ps1: line 369
369:         $results[2] | Should -BeGreaterThan 5
```

**Comment by PaulHigin** [comment]
@PoshChan  Please rerun tests

**Comment by PoshChan** [comment]
@PaulHigin, I do not understand the build target(s) `tests`; I only allow `static`, `windows`, `macos`, `linux`, `all`

**Comment by PaulHigin** [comment]
@PoshChan Please rerun all

**Comment by PoshChan** [comment]
@PaulHigin, successfully started rebuild of `PowerShell-CI-static-analysis, PowerShell-CI-Windows, PowerShell-CI-macOS, PowerShell-CI-Linux`

File: src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs (hunk 7 above)
Comment (PoshChan) [inline src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs]: @PaulHigin, your last commit had 1 failures in `PowerShell-CI-macos`
Test-Connection.Ping.Force IPv4 with implicit PingOptions
```powershell
Expected exactly '10.79.2.161', but got 10.79.0.26.
at <ScriptBlock>, /Users/runner/runners/2.165.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 83
83:             $result[0].Address | Should -BeExactly $realAddress
```

**Review by iSazonov — APPROVED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Core/Get-PSHostProcessInfo.Tests.ps1 [hunk 8]
```diff
@@ -53,4 +53,49 @@ Describe "Get-PSHostProcessInfo tests" -Tag CI {
         $psProcess.Count | Should -BeGreaterOrEqual 1
         $psProcess.ProcessId | Should -Contain $powershell.id
     }
+
+    It "Verifies named pipe filepath get method" {
+        $pipeFilePath = (Get-PSHostProcessInfo -Id $pid).GetPipeNameFilePath()
+        Test-Path -Path $pipeFilePath | Should -BeTrue
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Get-PSHostProcessInfo.Tests.ps1]:
```suggestion
        $pipeFilePath | Should -Exist
```

File: test/powershell/Modules/Microsoft.PowerShell.Core/Get-PSHostProcessInfo.Tests.ps1 [hunk 9]
```diff
@@ -53,4 +53,49 @@ Describe "Get-PSHostProcessInfo tests" -Tag CI {
...
+        $psFileName = $IsWindows ? 'pwsh.exe' : 'pwsh'
+        $psPath = Join-Path -Path $PSHOME -ChildPath $psFileName
+        $psProc = Start-Process -FilePath $psPath -ArgumentList "-File $testfilePath -LiveFilePath $aliveFile" -PassThru
+        Wait-UntilTrue -sb {
+            (Get-PSHostProcessInfo -Id $psProc.Id) -ne $null
+        } -TimeoutInMilliseconds 5000 -IntervalInMilliseconds 250
+
+        # Verify named pipe file path.
+        $psNamedPipePath = (Get-PSHostProcessInfo -Id $psProc.Id).GetPipeNameFilePath()
+        Test-Path -Path $psNamedPipePath | Should -BeTrue
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Get-PSHostProcessInfo.Tests.ps1]:
```suggestion
        $psNamedPipePath | Should -Exist
```

File: test/powershell/Modules/Microsoft.PowerShell.Core/Get-PSHostProcessInfo.Tests.ps1 [hunk 10]
```diff
@@ -53,4 +53,49 @@ Describe "Get-PSHostProcessInfo tests" -Tag CI {
...
+
+        # Signal PowerShell test process to exit normally.
+        Remove-Item -Path $aliveFile -Force -ErrorAction Ignore
+        Wait-UntilTrue -sb {
+            (Test-Path -Path $psNamedPipePath) -eq $false
+        } -TimeoutInMilliseconds 5000 -IntervalInMilliseconds 250
+
+        # Verify named pipe file path is removed.
+        Test-Path -Path $psNamedPipePath | Should -BeFalse
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Get-PSHostProcessInfo.Tests.ps1]:
```suggestion
        $psNamedPipePath | Should -Not -Exist
```

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Comment by PaulHigin** [comment]
@PoshChan  Please rerun all

**Comment by PoshChan** [comment]
@PaulHigin, successfully started rebuild of `PowerShell-CI-static-analysis, PowerShell-CI-Windows, PowerShell-CI-macOS, PowerShell-CI-Linux`

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by MaFreiberger** [comment]
PowerShell 7.1.0-preview.1 on debian 9.12

Problem still exists. Every time i power up a console (pwsh-preview) it creates something like
srwxr-xr-x  1 iteanova iteanova    0 Apr 21 15:07 CoreFxPipe_PSHost.D617DDBE.10883.None.pwsh
srwxr-xr-x  1 iteanova iteanova    0 Apr 21 15:08 CoreFxPipe_PSHost.D617DDE2.11055.None.pwsh
srwxr-xr-x  1 iteanova iteanova    0 Apr 21 15:09 CoreFxPipe_PSHost.D617DE06.11252.None.pwsh
srwxr-xr-x  1 iteanova iteanova    0 Apr 21 15:10 CoreFxPipe_PSHost.D617DE2A.11431.None.pwsh

**Comment by iSazonov** [comment]
@MaFreiberger It is not in Preview1. You can load nightly build or wait next Preview.

**Comment by rjmholt** [comment]
Should this be backported to PowerShell 7.0?

**Comment by iSazonov** [comment]
Have we requests for this?

**Comment by rjmholt** [comment]
Just labelled this as backport consider so that we can discuss whether this should be backported (since otherwise the LTS will just pollute named pipes onto the system). Just a suggestion though -- we can skip it if it's not thought to be severe.

**Pull request merged**